chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,263 @@
+		return err
+	}
+	sm.memberships[sm.currentEpoch+ConfigOffset+1] = mbs
+	return nil
+}
+
+// UpdateAndCheckVotes votes for the valSet and returns true if it has had enough votes for this valSet.
+func (sm *StateManager) UpdateAndCheckVotes(valSet *hierarchical.ValidatorSet) (bool, error) {
+	sm.membershipLock.Lock()
+	defer sm.membershipLock.Unlock()
+
+	h, err := valSet.Hash()
+	if err != nil {
+		return false, err
+	}
+	sm.reconfigurationVotes[string(h)]++
+	votes := int(sm.reconfigurationVotes[string(h)])
+	nodes := len(sm.memberships[sm.currentEpoch])
+	if votes < f(nodes)+1 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// applySnapshotRequest produces a StateSnapshotResponse event containing the current snapshot of the chat app state.
+// The snapshot is a binary representation of the application state that can be passed to applyRestoreState().
+func (sm *StateManager) applySnapshotRequest(snapshotRequest *eventpb.AppSnapshotRequest) (*events.EventList, error) {
+	sm.membershipLock.Lock()
+	defer sm.membershipLock.Unlock()
+
+	return events.ListOf(events.AppSnapshotResponse(
+		t.ModuleID(snapshotRequest.Module),
+		nil,
+		snapshotRequest.Origin,
+	)), nil
+}
+
+// applyRestoreState restores the application's state to the one represented by the passed argument.
+// The argument is a binary representation of the application state returned from Snapshot().
+func (sm *StateManager) applyRestoreState(snapshot *commonpb.StateSnapshot) (*events.EventList, error) {
+	sm.membershipLock.Lock()
+	defer sm.membershipLock.Unlock()
+
+	sm.currentEpoch = t.EpochNr(snapshot.Configuration.EpochNr)
+	sm.memberships = make([]map[t.NodeID]t.NodeAddress, len(snapshot.Configuration.Memberships))
+	for e, mem := range snapshot.Configuration.Memberships {
+		sm.memberships[e] = make(map[t.NodeID]t.NodeAddress)
+		for nID, nAddr := range mem.Membership {
+			var err error
+			sm.memberships[e][t.NodeID(nID)], err = multiaddr.NewMultiaddr(nAddr)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	newMembership := copyMap(sm.memberships[snapshot.Configuration.EpochNr+ConfigOffset])
+	sm.memberships = append(sm.memberships, newMembership)
+
+	return events.EmptyList(), nil
+}
+
+func (sm *StateManager) OrderedValidatorsAddresses() []t.NodeID {
+	sm.membershipLock.Lock()
+	defer sm.membershipLock.Unlock()
+
+	membership := sm.memberships[sm.currentEpoch]
+	sortedIDs := getSortedKeys(membership)
+	return sortedIDs
+}
+
+// The ImplementsModule method only serves the purpose of indicating that this is a Module and must not be called.
+func (sm *StateManager) ImplementsModule() {}
+
+func getSortedKeys(m map[t.NodeID]t.NodeAddress) []t.NodeID {
+	skeys := make([]string, len(m))
+	i := 0
+	for k := range m {
+		skeys[i] = k.Pb()
+		i++
+	}
+
+	sort.Strings(skeys)
+
+	keys := make([]t.NodeID, len(m))
+	for k := range skeys {
+		keys[k] = t.NodeID(skeys[k])
+	}
+
+	return keys
+}
+
+func copyMap(m map[t.NodeID]t.NodeAddress) map[t.NodeID]t.NodeAddress {
```

Comment:
you can use `maputil.Copy`

Reply:
Eudico supports Go 1.17. The project is not built with features of Go 1.18+

Reply:
That's true, but this part of the code is Mir-specific, and Mir needs Go 1.18+ anyway, so not using generics here doesn't seem of any benefit.

That said, it's a very minor, mostly aesthetic issue, so feel free to keep this separate implementation if you prefer and resolve this comment yourself.

Reply:
I cannot do that without changing Go version for eudico/Lotus
```
chain/consensus/mir/state_manager.go:210:35: implicit function instantiation requires go1.18 or later (-lang was set to go1.17; check go.mod)
```

Reply:
Ah ok, then of course let's keep it.